### PR TITLE
21 06 07

### PR DESCRIPTION
**Pull request by lindajohnstone**

### CODE REVIEW

File: ConwaysGameOfLife.Tests/LocationShould.cs [hunk 1]
```diff
@@ -57,5 +29,61 @@ public void WhenTwoLocationsInstantiated_NotBeEqual(int x1, int y1, int x2, int
...
+                new Location(2, 0),
+                new Location(2, 1),
+                new Location(2, 2)
+            };
+
+            Assert.True(UniverseHelper.ListsOfLocationsAreEqual(expected, result));
+        }
+
+        [Fact]
+        public void WhenTwoListsOfCellsNotSameOrderInstantiated_BeEqual()
```
Comment (indianbread) [inline ConwaysGameOfLife.Tests/LocationShould.cs:62]: the naming of this should be list of locations as above

File: ConwaysGameOfLife.Tests/RulesShould.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+using System;
+using Moq;
+using Xunit;
+
+namespace ConwaysGameOfLife.Tests
+{
+    public class RulesShould
+    {
+        [Fact]
+        public void WhenAnyLiveCellHasFewerThan2LiveNeighbours_ItDies()
```
Comment (indianbread) [inline ConwaysGameOfLife.Tests/RulesShould.cs:10]: keen to hear your thought process behind how this will work

File: ConwaysGameOfLife.Tests/UniverseHelper.cs [hunk 3]
```diff
@@ -33,35 +33,35 @@ public static bool UniversesAreEqual(Universe obj1, Universe obj2)
...
         {
-            var isLocationSame = (obj1.Location.X == obj2.Location.X) && (obj1.Location.Y == obj2.Location.Y);
-            var isCellStateSame = obj1.CellState == obj2.CellState;
+            var isLocationSame = (cell1.Location.X == cell2.Location.X) && (cell1.Location.Y == cell2.Location.Y);
+            var isCellStateSame = cell1.CellState == cell2.CellState;
             return isCellStateSame && isLocationSame;
         }
 
-        public static bool ListsOfCellsAreEqual(List<Cell> obj1, List<Cell> obj2)
+        public static bool ListsOfLocationsAreEqual(List<Location> location1, List<Location> location2)
```
Comment (indianbread) [inline ConwaysGameOfLife.Tests/UniverseHelper.cs:43]: suggestion: use plural for a variable associated with a collection (list)

File: ConwaysGameOfLife.Tests/UniverseHelper.cs [hunk 4]
```diff
@@ -33,35 +33,35 @@ public static bool UniversesAreEqual(Universe obj1, Universe obj2)
...
             {
                 return false;
             }
-            var obj1Ordered = obj1.OrderBy(c => c.Location.X).ThenBy(c => c.Location.Y).ToList();
-            var obj2Ordered = obj2.OrderBy(c => c.Location.X).ThenBy(c => c.Location.Y).ToList();
+            var location1Ordered = location1.OrderBy(c => c.X).ThenBy(c => c.Y).ToList();
+            var location2Ordered = location2.OrderBy(c => c.X).ThenBy(c => c.Y).ToList();
 
-            for (var cell = 0; cell < obj1.Count; cell++)
+            for (var location = 0; location < location1.Count; location++)
```
Comment (indianbread) [inline ConwaysGameOfLife.Tests/UniverseHelper.cs:52]: suggestion: rename location to i as it's an index and not a location

File: ConwaysGameOfLife.Tests/UniverseShould.cs [hunk 5]
```diff
@@ -20,7 +20,7 @@ public void WhenInitialized_BePopulatedWithAllDeadCells()
         }  
 
         [Fact]
-        public void WhenUniverseGenerated_ReturnsListOfNeighbours()
+        public void WhenGivenANonBoundaryCell_ReturnCorrectNeighbours()
```
Comment (indianbread) [inline ConwaysGameOfLife.Tests/UniverseShould.cs:23]: nice naming

File: ConwaysGameOfLife.Tests/UniverseShould.cs [hunk 6]
```diff
@@ -134,7 +134,7 @@ public void WhenTwoUniversesInstantiated_BeEqual()
         }
 
         [Fact]
-        public void WhenTwoUniversesInstantiated_NotBeEqual()
+        public void WhenUniversesOfDifferentDimensionsInstantiated_NotBeEqual()
```
Comment (indianbread) [inline ConwaysGameOfLife.Tests/UniverseShould.cs:137]: the test is not doing what the name is suggesting.

File: ConwaysGameOfLife.Tests/UniverseShould.cs [hunk 7]
```diff
@@ -154,9 +154,113 @@ public void WhenTwoUniversesInstantiated_NotBeEqual()
...
+                new Location(1, 2)
+            };
+
+            var result = universe.GetCellNeighbourLocations(cell);
+
+            Assert.True(UniverseHelper.ListsOfLocationsAreEqual(expected, result));
+        }
+
+        [Fact]
+        public void WhenGivenABoundaryCellX_ReturnCorrectNeighbourLocationsListOutOfOrder()
```
Comment (indianbread) [inline ConwaysGameOfLife.Tests/UniverseShould.cs:204]: what is the purpose of this test compared to the previous one? Is it required?

File: ConwaysGameOfLife.Tests/UniverseShould.cs [hunk 8]
```diff
@@ -154,9 +154,113 @@ public void WhenTwoUniversesInstantiated_NotBeEqual()
...
+                new Location(1, 2)
+            };
+
+            var result = universe.GetCellNeighbourLocations(cell);
+
+            Assert.True(UniverseHelper.ListsOfLocationsAreEqual(expected, result));
+        }
+
+        [Fact]
+        public void WhenGivenABoundaryCellY_ReturnCorrectNeighbourLocationsListOutOfOrder()
```
Comment (indianbread) [inline ConwaysGameOfLife.Tests/UniverseShould.cs:225]: does the order of the locations matter, as long as they're correct?

File: ConwaysGameOfLife/Universe.cs [hunk 9]
```diff
@@ -48,15 +48,21 @@ public bool AreAllCellsDead()
 
         public List<Location> GetCellNeighbourLocations(Cell cell)
         {
+            var xMiddle = cell.Location.X;
```
Comment (indianbread) [inline ConwaysGameOfLife/Universe.cs:51]: suggestion: maybe call this plain x and y

File: ConwaysGameOfLife/Universe.cs [hunk 10]
```diff
@@ -65,7 +71,7 @@ public List<Location> GetCellNeighbourLocations(Cell cell)
         public int CountLiveNeighbours(Cell cell)
         {
             var aliveNeighbours = new List<Location>();
```
Comment (indianbread) [inline ConwaysGameOfLife/Universe.cs:73]: alternative: `var aliveNeighbours = 0;`
then do the check in the loop and increment by 1 if the cell is alive

**Review by indianbread — APPROVED** [review]
good work setting up the neighbours & checking the boundary conditions!

File: ConwaysGameOfLife.Tests/LocationShould.cs [hunk 11]
```diff
@@ -1,41 +1,13 @@
...
-            var one = location;
-            var two = new Location(0, 0);
-
-            var result = (one == two);
-
-            Assert.True(result);
-        }
-
         [Theory]
         [InlineData(0, 0, 0, 0, true)]
```
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/LocationShould.cs:10]: Can we add an extra case here, remove the `expected` param, and `Assert.True` instead
```suggestion
        [InlineData(0, 0, 0, 0)]
        [InlineData(5, 5, 5, 5)]
```

Comment (kedoodle) [reply ConwaysGameOfLife.Tests/LocationShould.cs:10]: Same goes for the `WhenTwoLocationsInstantiated_NotBeEqual` test, but `Assert.False`

File: ConwaysGameOfLife.Tests/LocationShould.cs [hunk 12]
```diff
@@ -57,5 +29,61 @@ public void WhenTwoLocationsInstantiated_NotBeEqual(int x1, int y1, int x2, int
 
             Assert.Equal(expected, result);
         }
+
+        [Fact]
+        public void WhenTwoListsOfLocationsSameOrderInstantiated_BeEqual()
```
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/LocationShould.cs:34]: No need to say `Instantiated` as we're not really testing the constructor for a class you've written
```suggestion
        public void GivenTwoListsOfLocationsSameOrder_BeEqual()
```

File: ConwaysGameOfLife.Tests/LocationShould.cs (hunk 1 above)
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/LocationShould.cs:62]: Same goes re: `Instantiated`
```suggestion
        public void GivenTwoListsOfLocationsNotSameOrder_BeEqual()
```

File: ConwaysGameOfLife.Tests/LocationShould.cs [hunk 13]
```diff
@@ -1,41 +1,13 @@
...
-
-            var result = (one == two);
-
-            Assert.True(result);
-        }
-
         [Theory]
         [InlineData(0, 0, 0, 0, true)]
-        [InlineData(0, 0, 1, 0, false)]
         public void WhenTwoLocationsInstantiated_BeEqual(int x1, int y1, int x2, int y2, bool expected)
```
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/LocationShould.cs:11]: No need to say Instantiated as we're not really testing the constructor for a class you've written

```suggestion
        public void GivenTwoLocationsOfSameCoordinates_BeEqual(int x1, int y1, int x2, int y2)
```

Comment (kedoodle) [reply ConwaysGameOfLife.Tests/LocationShould.cs:11]: Same goes for the `WhenTwoLocationsInstantiated_NotBeEqual `test  i.e. could be renamed to `GivenTwoLocationsOfDifferingCoordinates_NotBeEqual`

File: ConwaysGameOfLife.Tests/RulesShould.cs [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+namespace ConwaysGameOfLife.Tests
+{
+    public class RulesShould
+    {
+        [Fact]
+        public void WhenAnyLiveCellHasFewerThan2LiveNeighbours_ItDies()
+        {
+            var universe = new Universe(3, 3);
+            var rules = new Rules(universe);
+            var cell = new Cell(It.IsAny<CellState>(), 1, 1);
```
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/RulesShould.cs:14]: Do we need a `Cell` or just a `Location` here? What is the purpose of stubbing the `CellState`?

File: ConwaysGameOfLife.Tests/UniverseHelper.cs [hunk 15]
```diff
@@ -6,25 +6,25 @@ namespace ConwaysGameOfLife.Tests
 {
     public static class UniverseHelper
     {
-        public static bool UniversesAreEqual(Universe obj1, Universe obj2)
+        public static bool UniversesAreEqual(Universe universe1, Universe universe2)
         {
-            var isSizeSame = (obj1.GridWidth == obj2.GridWidth) && (obj1.GridLength == obj2.GridLength);
+            var isSizeSame = (universe1.GridWidth == universe2.GridWidth) && (universe1.GridLength == universe2.GridLength);
```
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/UniverseHelper.cs:11]: Extra consideration: this is actually only checking the grid dimensions - what about if the actual number of `Cells` were to be different? Could be an extra check early on which is just as simple as this existing one

File: ConwaysGameOfLife.Tests/UniverseHelper.cs (hunk 3 above)
Comment (kedoodle) [reply ConwaysGameOfLife.Tests/UniverseHelper.cs:43]: either `locations1` or `locationList1` works

File: ConwaysGameOfLife.Tests/UniverseHelper.cs [hunk 16]
```diff
@@ -33,35 +33,35 @@ public static bool UniversesAreEqual(Universe obj1, Universe obj2)
...
                 }
             }
             return true;
         }
 
-        public static bool LocationsAreEqual(Location obj1, Location obj2)
+        public static bool LocationsAreEqual(Location location1, Location location2)
         {
-            if (obj1.X == obj2.X && obj1.Y == obj2.Y)
+            if (location1.X == location2.X && location1.Y == location2.Y)
```
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/UniverseHelper.cs:64]: can just return this like 
```suggestion
            return location1.X == location2.X && location1.Y == location2.Y;
```

File: ConwaysGameOfLife.Tests/UniverseShould.cs [hunk 17]
```diff
@@ -37,7 +37,7 @@ public void WhenUniverseGenerated_ReturnsListOfNeighbours()
 
             var result = universe.GetCellNeighbourLocations(cell);
 
-            Assert.Equal(expected, result);
+            Assert.True(UniverseHelper.ListsOfLocationsAreEqual(expected, result));
             Assert.Equal(8, result.Count);
```
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/UniverseShould.cs:41]: the count equality should be encapsulated in the first assertion
```suggestion
            Assert.True(UniverseHelper.ListsOfLocationsAreEqual(expected, result));
```

File: ConwaysGameOfLife.Tests/LocationShould.cs [hunk 18]
```diff
@@ -57,5 +29,61 @@ public void WhenTwoLocationsInstantiated_NotBeEqual(int x1, int y1, int x2, int
...
+                new Location(1, 0),
+                new Location(2, 0),
+                new Location(2, 1),
+                new Location(2, 2)
+            };
+
+            Assert.True(UniverseHelper.ListsOfLocationsAreEqual(expected, result));
+        }
     }
 }
```
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/LocationShould.cs:89]:
```suggestion
}

```

File: ConwaysGameOfLife.Tests/RulesShould.cs [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+
+            var result = rules.Check(cell);
+
+            Assert.True(result);
+        }
+    }
+}
```
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/RulesShould.cs:21]:
```suggestion
}

```

File: ConwaysGameOfLife.Tests/UniverseShould.cs (hunk 6 above)
Comment (kedoodle) [reply ConwaysGameOfLife.Tests/UniverseShould.cs:137]: could be a sign that the helper method isn't doing as it should i.e.
- shouldn't throw exceptions
- should always return true or false

File: ConwaysGameOfLife.Tests/UniverseShould.cs (hunk 7 above)
Comment (kedoodle) [reply ConwaysGameOfLife.Tests/UniverseShould.cs:204]: Sorry I may have suggested that it's okay to have the expected list in a different order without clarifying that the `UniverseHelper.ListsOfLocationsAreEqual` should already handle when the lists aren't in the same order

File: ConwaysGameOfLife.Tests/UniverseShould.cs [hunk 20]
```diff
@@ -154,9 +154,113 @@ public void WhenTwoUniversesInstantiated_NotBeEqual()
...
+                new Location(1, 1),
+                new Location(1, 2)
+            };
+
+            var result = universe.GetCellNeighbourLocations(cell);
+
+            Assert.True(UniverseHelper.ListsOfLocationsAreEqual(expected, result));
         }
     }
 }
```
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/UniverseShould.cs:266]:
```suggestion
}

```

File: ConwaysGameOfLife.Tests/UniverseShould.cs [hunk 21]
```diff
@@ -154,9 +154,113 @@ public void WhenTwoUniversesInstantiated_NotBeEqual()
...
+                new Location(1, 2)
+            };
+
+            var result = universe.GetCellNeighbourLocations(cell);
+
+            Assert.True(UniverseHelper.ListsOfLocationsAreEqual(expected, result));
+        }
+
+        [Fact]
+        public void WhenGivenABoundaryCornerCell_ReturnCorrectNeighbourLocationsListOutOfOrder()
```
Comment (kedoodle) [inline ConwaysGameOfLife.Tests/UniverseShould.cs:246]: love that you're testing several "kinds" of locations i.e.
- non boundary
- single boundary
- dual boundary

could think about extending this with `[Theory]` to become:
- non boundary
- single boundary on y axis (case each for top and bottom)
- single boundary on x axis (case each for left and right)
- all four corners?

File: ConwaysGameOfLife/Location.cs [hunk 22]
```diff
@@ -15,52 +15,5 @@ public Location(int x, int y)
...
-        public static bool operator !=(Location obj1, Location obj2)
-        {
-            if (!OperatorOverride(obj1, obj2))
-            {
-                return false;
-            }
-            return true;
-        }
     }
 }
```
Comment (kedoodle) [inline ConwaysGameOfLife/Location.cs:19]:
```suggestion
}

```

File: ConwaysGameOfLife/OutputFormatter.cs [hunk 23]
```diff
@@ -24,7 +24,7 @@ public string FormatWelcomeMessage()
             return "Welcome to Game of Life!";
         }
 
-        public string FormatUniverse()
+        public string FormatUniverse() // universe created with user input - dimensions & live cells ?? - needed??
```
Comment (kedoodle) [inline ConwaysGameOfLife/OutputFormatter.cs:27]: we can consider writing some tests for this formatter even though we haven't started actually outputting anything to the console

```suggestion
        public string FormatUniverse(Universe universe) // universe created with user input - dimensions & live cells ?? - needed??
```

File: ConwaysGameOfLife/Rules.cs [hunk 24]
```diff
@@ -8,6 +10,13 @@ public class Rules
         // * Any live cell with more than three live neighbours dies, as if by overcrowding.
         // * Any live cell with two or three live neighbours lives on to the next generation.
         // * Any dead cell with exactly three live neighbours becomes a live cell.
-        
+        public Rules(Universe universe)
```
Comment (kedoodle) [inline ConwaysGameOfLife/Rules.cs:13]: Which way should this dependency be? We're currently injecting a `Universe` into an instance of `Rules`

File: ConwaysGameOfLife/Universe.cs (hunk 9 above)
Comment (kedoodle) [reply ConwaysGameOfLife/Universe.cs:51]: we paired on this, I mentioned `xMiddle` could be more explicit than `x` if you want to avoid potential confusion. I have no preference personally

File: ConwaysGameOfLife/Universe.cs (hunk 10 above)
Comment (kedoodle) [reply ConwaysGameOfLife/Universe.cs:73]: can also rejig this function (with some more widespread refactoring) to do a Linq count. we can discuss this tomorrow

File: ConwaysGameOfLife/Universe.cs [hunk 25]
```diff
@@ -65,7 +71,7 @@ public List<Location> GetCellNeighbourLocations(Cell cell)
         public int CountLiveNeighbours(Cell cell)
         {
             var aliveNeighbours = new List<Location>();
-            ThrowsExceptionIfCellIsNull(cell);
```
Comment (kedoodle) [inline ConwaysGameOfLife/Universe.cs:68]: I'm not a fan of throwing random unhandled exceptions. There are more graceful ways of handling stuff like this e.g. simply not passing any potential nulls to `CountLiveNeighbours`

File: ConwaysGameOfLife/Rules.cs [hunk 26]
```diff
@@ -8,6 +10,13 @@ public class Rules
...
+        {
+        }
+
+        public bool Check(Cell cell)
+        {
+            throw new NotImplementedException();
+        }
     }
 }
```
Comment (kedoodle) [inline ConwaysGameOfLife/Rules.cs:22]:
```suggestion
}

```

**Review by kedoodle — APPROVED** [review]
a few nitpicks and improvements. nice work on getting the neighbour tests and function going, as well as starting on the rules.

there's an opportunity to do a bunch of refactoring once the neighbour stuff is finalised - have a think about what calls will actually be made on the `Universe` eventually (hint: this could even be none! or it could be a fraction of what public functions are currently available)
